go-update/src/main/java/org/reactome/release/goupdate/GoTermsReconciler.java

```diff
@@ -0,0 +1,200 @@
+package org.reactome.release.goupdate;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.gk.model.GKInstance;
+import org.gk.model.ReactomeJavaConstants;
+import org.gk.persistence.MySQLAdaptor;
+import org.gk.schema.InvalidAttributeException;
+
+/**
+ * This class should be used to reconcile between the GO file and the database, after updates have been attempted.
+ * @author sshorser
+ *
+ */
+public class GoTermsReconciler
+{
+
+	private static final Logger reconciliationLogger = LogManager.getLogger("reconciliationLog");
+	private MySQLAdaptor adaptor;
+	
+	public GoTermsReconciler(MySQLAdaptor adaptor)
+	{
+		this.adaptor = adaptor;
+	}
+	
+	/**
+	 * Attempts to reconcile between the database and the terms from the file. Reconciliation reports are logged to a file (not returned).
+	 * @param goTermsFromFile - GO terms from the file.
+	 * @param goToECNumbers - GO-to-EC Numbers, from the file.
+	 * @throws Exception
+	 */
+	public void reconcile(Map<String, Map<String, Object>> goTermsFromFile, Map<String, List<String>> goToECNumbers) throws Exception
+	{
+		for (String goAccession : goTermsFromFile.keySet())
+		{
+			Map<String, Object> goTerm = goTermsFromFile.get(goAccession);
+			@SuppressWarnings("unchecked")
+			Collection<GKInstance> instances = this.adaptor.fetchInstanceByAttribute( ((GONamespace)goTerm.get(GoUpdateConstants.NAMESPACE)).getReactomeName(), ReactomeJavaConstants.accession, "=", goAccession );
+			if (instances != null)
+			{
+				if (instances.size()>1)
+				{
+					reconciliationLogger.warn("GO Accession {} appears {} times in the database. It should probably only appear once.",goAccession, instances.size());
+				}
+				for (GKInstance instance : instances)
+				{
+					this.adaptor.fastLoadInstanceAttributeValues(instance);
+					// We'll just grab all relationships in advance.
+					Collection<GKInstance> instancesOfs = new ArrayList<>();
+					Collection<GKInstance> partOfs = new ArrayList<>();
+					Collection<GKInstance> hasParts = new ArrayList<>();
+					if (instance.getSchemClass().isa(ReactomeJavaConstants.GO_CellularComponent))
+					{
+						instancesOfs = (Collection<GKInstance>) instance.getAttributeValuesList(ReactomeJavaConstants.instanceOf);
+						partOfs = (Collection<GKInstance>) instance.getAttributeValuesList(ReactomeJavaConstants.componentOf);
+						hasParts = (Collection<GKInstance>) instance.getAttributeValuesList("hasPart");
+					}
+
+					for (String k : goTerm.keySet())
+					{
+						switch (k)
+						{
+							case GoUpdateConstants.DEF:
```

Comment:
Comments for each of these cases, I'm not really sure what it means to be a 'DEF'

Reply:
DEF is GO's shorthand, I think, for definition.  Maybe the value that the constant holds could be as is, but the constant name could be "DEFINITION"?

Reply:
I'd prefer to keep it consistent with GO's terminology, in this case.

Reply:
Maybe a comment to clarify the abbreviation then?

Reply:
Javadoc has been added to the appropriate constant to explain the meaning of DEF.